### ISSUE TITLE
Cannot run multiple instances of bosh in parallel with a singularity image

### ISSUE THREAD

**Issue by prioux** [boutiques/boutiques#351]
If invoking bosh multiple time on the same descriptor in the same directory, and the descriptor defines an execution environment in a singularity image, each bosh will attempt to pull and install the singularity image on top of each other. E.g. assuming I want to process some dataset with three different set of parameters:

```shell
bosh exec launch descriptor.json invoke_params1.json &
bosh exec launch descriptor.json invoke_params2.json &
bosh exec launch descriptor.json invoke_params3.json &
```

and assuming descriptor.json has a definition like

```json
    "container-image": {
        "image": "bids/example",
        "index": "docker://",
        "type": "singularity"
    },
```

then all three instances will try to write simultaneously to the file `bids-example.simg` and of course everything will crash and burn and explode and wail in pain.

Two possible solutions:

1) build into bosh the ability to recognize that another bosh is pulling the image and wait for it:

The pull would be performed on a temporary but deterministic name, e.g. `bids-example.simg-downloading` and then renamed at the end, so that other boshes can wait for it.

2) provide a bosh command to separately pull the image and return its path:

```shell
bosh exec prepare descriptor.json
image=bids-example.simg
```

so that we can parse the output of `prepare` and pass the image with `bosh blah blah --simg image`

**Comment by glatard** [comment]
Thanks for reporting this, I think solution 2 is better as it's simpler and the command could be used in other contexts too.

**Comment by prioux** [comment]
It would be even nicer if we could control the location of the image downloaded:

```shell
bosh exec prepare descriptor.json invoke.json --imagepath=/my/cache/location/hello.simg
```

**Comment by glatard** [comment]
Additional notes:

If the image is a Docker image `prepare` should just pull it, so that multiple instances don't step on each other.

`prepare` should be idempotent, i.e., we should be able to run it sequentially as many times as we want without side effects. That way, `bosh exec launch` could invoke `prepare` in place of the current pulling mechanism without worrying about triggering additional transfers.

It would also be nice if `prepare` could be thread-safe to simplify its use: a simple lockfile should do. That would also enable both @prioux's solutions 1. and 2.

**Comment by prioux** [comment]
Yep, all good.

**Comment by glatard** [comment]
Implemented by @erinb90 in just-released 0.5.16 version.
